Refactor the repository layout

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

> [!NOTE]
> This is built on top of #376 because we need the updated version of [cisagov/pre-commit-packer]. The changes for just this pull request can be viewed at https://github.com/cisagov/skeleton-packer/compare/lineage/skeleton...improvement/refactor_repository

This pull request reformats the repository layout including breaking the Packer template into multiple files.
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

With the `0.1.0` release of [cisagov/pre-commit-packer] we can now use Packer templates that consist of more than one file. Breaking up this template will make it easier to maintain and modify. As part of that it makes sense to treat Packer projects more like Terraform root modules since both Packer and Terraform are Hashicorp projects that use the Hashicorp configuration language.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

Automated tests pass.
<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

[cisagov/pre-commit-packer]: https://github.com/cisagov/pre-commit-packer

mcdonnnj committed:
Add a pre-commit hook to run pip-audit

The pip-audit tool will audit any supplied pip requirements files for
vulnerable packages.

Commit:
Add checks for semantic python versions

Commit:
Add checks for semantic python versions

Commit:
Refactor code for the semantic check

This commit will make a few changes. The
orginal version of the semantic checking
function was a bit more difficult to read.
It is now somewhat easier to follow how
the regex is structured. Also the function
has been renamed to check_python_version
since it has 2 functions, making sure that
the version is semantically correct and the
second is to make sure that it is installed
on the user's machine. This makes it easier
to follow the logic for the flags, -p or
--python-version and -l or --list-versions

Commit:
Refactor code for the semantic check

This commit will make a few changes. The
orginal version of the semantic checking
function was a bit more difficult to read.
It is now somewhat easier to follow how
the regex is structured. Also the function
has been renamed to check_python_version
since it has 2 functions, making sure that
the version is semantically correct and the
second is to make sure that it is installed
on the user's machine. This makes it easier
to follow the logic for the flags, -p or
--python-version and -l or --list-versions

Commit:
Merge branch 'improvement/correct-semantic-python-version-checks' of https://github.com/cisagov/skeleton-generic into improvement/correct-semantic-python-version-checks

Commit:
Remove example of correct semantic version

Commit:
Refactor the error message for the user

Commit:
Improve the semantic error message

michaelsaki committed:
Fix grammar

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

Commit:
Refactor regex, add link, and improve comments

Commit:
Update link to use semver.org over regex101.com

michaelsaki committed:
Remove unnecessary period

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj committed:
Add a meta hook to the pre-commit configuration

Add the `check-useless-excludes` meta hook to verify that any defined
`exclude` directives apply to at least one file in the repository.

mcdonnnj committed:
Remove `exclude` directive that does not apply to any files

jsf9k committed:
Add a lower-bound pin for flake8-docstrings

mcdonnnj committed:
Use the hashicorp/setup-packer GitHub Action

Instead of manually installing Packer we can instead leverage the
hashicorp/setup-packer Action just as we do for Terraform.

mcdonnnj committed:
Update pre-commit hook versions

This is done automatically with the `pre-commit autoupdate` command.
The pre-commit/mirrors-prettier hook was manually held back because the
latest tags are for alpha releases of the next major version.

mcdonnnj committed:
Manually update the prettier hook

Use the latest v3 release available from NPM.

mcdonnnj committed:
Use the rbubley/mirrors-prettier hook for prettier

This replaces the now archived pre-commit/mirrors-prettier hook.

mcdonnnj committed:
Add a new trigger for the sync-labels GitHub Actions workflow

Add a `workflow_dispatch` trigger so we can manually run the workflow
if needed.

mcdonnnj committed:
Remove unnecessary quotes in the sync-labels workflow

jsf9k committed:
Remove @jasonodoom as a codeowner

He is no longer a member of @cisagov/vm-dev.

jsf9k committed:
Pin to a specific version

Previously we only provided a lower bound for the version, but pinning to a specific version aligns with what has been done with the prettier hook and how pre-commit hooks are pinned in general.

The flake8-docstrings package is rarely updated, so there is no real downside to pinning to a specific version.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj committed:
Add four new hooks from pre-commit/pre-commit-hooks

dependabot[bot] committed:
Bump actions/cache from 3 to 4

Bumps [actions/cache](https://github.com/actions/cache) from 3 to 4.
- [Release notes](https://github.com/actions/cache/releases)
- [Changelog](https://github.com/actions/cache/blob/main/RELEASES.md)
- [Commits](https://github.com/actions/cache/compare/v3...v4)

---
updated-dependencies:
- dependency-name: actions/cache
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump crazy-max/ghaction-github-status from 3 to 4

Bumps [crazy-max/ghaction-github-status](https://github.com/crazy-max/ghaction-github-status) from 3 to 4.
- [Release notes](https://github.com/crazy-max/ghaction-github-status/releases)
- [Commits](https://github.com/crazy-max/ghaction-github-status/compare/v3...v4)

---
updated-dependencies:
- dependency-name: crazy-max/ghaction-github-status
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

mcdonnnj committed:
Merge pull request #170 from cisagov/dependabot/github_actions/crazy-max/ghaction-github-status-4

Bump crazy-max/ghaction-github-status from 3 to 4

mcdonnnj committed:
Merge pull request #171 from cisagov/dependabot/github_actions/actions/cache-4

Bump actions/cache from 3 to 4

mcdonnnj committed:
Merge pull request #187 from cisagov/improvement/use_setup_packer_action

Use an Action to install Packer in our GitHub Actions workflows

mcdonnnj committed:
Merge pull request #176 from cisagov/improvement/correct-semantic-python-version-checks

Add checks for correct semantic version of Python

mcdonnnj committed:
Merge pull request #188 from cisagov/remove-odoom-as-a-codeowner

Remove @jasonodoom as a codeowner

mcdonnnj committed:
Merge pull request #178 from cisagov/improvement/add_pre-commit_meta_hook

Add the `check-useless-excludes` hook to the pre-commit configuration

mcdonnnj committed:
Merge pull request #179 from cisagov/improvement/add_pip-audit_pre-commit_hook

Add a pre-commit hook to run `pip-audit`

mcdonnnj committed:
Merge pull request #184 from cisagov/improvement/pin-flake8-docstrings

Add a lower-bound pin for the `flake8-docstrings` `pip` package

mcdonnnj committed:
Merge pull request #185 from cisagov/maintenance/update_pre-commit_hooks

Update `pre-commit` hooks

mcdonnnj committed:
Merge pull request #186 from cisagov/improvement/switch_prettier_hook

Switch the `pre-commit` hook used to run `prettier`

mcdonnnj committed:
Move the Packer template to the root of the repository

mcdonnnj committed:
Move the version file to the root of the repository

mcdonnnj committed:
Rename src/ to ansible/

This aligns with what the src/ directory still contains.

mcdonnnj committed:
Move the Packer template variables to their own file

This also requires us to replace any Packer commands that reference the
`packer.pkr.hcl` file to instead reference the directory that houses
the template (the root directory `.`).

mcdonnnj committed:
Move the Packer template locals to their own file

mcdonnnj committed:
Move the Packer template's build block to its own file

mcdonnnj committed:
Move the Packer template's AMI sources to their own files

mcdonnnj committed:
Move the Packer template's AMI data blocks to their own file

mcdonnnj committed:
Update the README

Fix the link for variables in the template and update all references to
a Packer configuration to refer to a Packer template instead.

mcdonnnj committed:
Add a `terraform-docs` configuration

Add the `terraform-docs` configuration file and markers for where to
insert output in the README.

mcdonnnj committed:
Run `terraform-docs` to populate the README

mcdonnnj committed:
Update the README

Now that we use `terraform-docs` to populate information about the
template's variables we do not need a manually managed table with that
information in the README.

mcdonnnj committed:
Rename the `packer.pkr.hcl` file

Calling it `versions.pkr.hcl` aligns with how we name files in our
Terraform projects.

mcdonnnj committed:
Simply the CODEOWNERS configuration for the Packer template

There is no reason to list each file if we can wildcard match on the
file extension.

mcdonnnj committed:
Update pre-commit hook versions

This is done automatically with the `pre-commit autoupdate` command.

jsf9k commented on README.md:
```diff
@@ -129,12 +117,12 @@ Here is an example of how to kick off a pre-release build:
 
 If you are satisfied with your pre-release image, you can easily create a release
-that deploys to all regions by adding additional regions to the packer configuration.
+that deploys to all regions by adding additional regions to the Packertemplate.
```
```suggestion
that deploys to all regions by adding additional regions to the Packer template.
```

jsf9k commented on variables.pkr.hcl:
Should we go ahead and add `nullable = false` for all these variables?

mcdonnnj replied:
I don't believe that is a valid [argument for Packer input variables](https://developer.hashicorp.com/packer/docs/templates/hcl_templates/variables#arguments). It's specifically something for Terraform modules as far as I can tell.

dav3r reviewed (approved):
👍 👍

mcdonnnj committed:
Add a missed space in the README

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

mcdonnnj committed:
Restrict permissions of GITHUB_TOKEN

This changes the default permissions for the GITHUB_TOKEN used in our
GitHub Actions configuration to the minimum required to successfully
run.

mcdonnnj committed:
Add the GitHubSecurityLab/actions-permissions/monitor Action

This Action will provide information about the usage of GITHUB_TOKEN in
the workflow. It should be added to _every_ job in _any_ workflow to
provide information for analysis.

mcdonnnj committed:
Sort hook ids in each pre-commit hook entry

Ensure that all hook ids are sorted alphabetically in each hook entry
in our pre-commit configuration.

mcdonnnj committed:
Merge pull request #189 from cisagov/improvement/manually_run_sync-labels_workflow

Allow the `sync-labels` workflow to be run manually

mcdonnnj committed:
Merge pull request #190 from cisagov/improvement/add_actions-permissions-monitor

Add the `GitHubSecurityLab/actions-permissions/monitor` Action

mcdonnnj committed:
Merge pull request #191 from cisagov/improvement/github_tokenn_polp

Explicitly define permissions of `GITHUB_TOKEN` in our GitHub Actions workflows

mcdonnnj committed:
Merge pull request #192 from cisagov/maintenance/update_pre-commit_hooks

Update `pre-commit` hook versions

mcdonnnj committed:
Merge pull request #193 from cisagov/improvement/add_more_pre-commit_hooks

Add additional hooks from `pre-commit/pre-commit-hooks`

mcdonnnj committed:
Merge pull request #194 from cisagov/improvement/ensure_pre-commit_hooks_are_sorted

Sort hook ids in each `pre-commit` hook entry

jsf9k committed:
Merge remote-tracking branch 'skeleton-generic/develop' into lineage/skeleton

jsf9k committed:
Uncomment new Dependabot ignore directive from upstream

jsf9k committed:
Install Packer via hashicorp/setup-packer everywhere

jsf9k committed:
Upgrade Bandit to 1.7.10

We should use the same version of Bandit throughout the pre-commit
configuration.

jsf9k committed:
Add explicit permissions for jobs that lack them

jsf9k committed:
Add GitHubSecurityLab/actions-permissions/monitor task to each job that lacks it

jsf9k committed:
Remove shebang

This file is not executable and hence it should not have a shebang.

jsf9k committed:
Add repo read permissions for all jobs that checkout code

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Remove repeated comment

mcdonnnj committed:
Update the commented out dependabot ignore directives

Add a directive for hashicorp/setup-packer that was missed when it was
added to the `build` workflow. Add a directive for
cisagov/setup-env-github-action that is not strictly necessary since we
currently just pull from the `develop` branch, but is good to have in
case we were to change that in the future.

mcdonnnj committed:
Merge pull request #195 from cisagov/bug/add_missing_dependabot_ignore

Add missing dependabot ignore directives

jsf9k committed:
Uncomment new Dependabot directives from upstream

jsf9k committed:
Update the commented out dependabot ignore directives

Add a directive for hashicorp/setup-packer that was missed when it was
added to the `build` workflow. Add a directive for
cisagov/setup-env-github-action that is not strictly necessary since we
currently just pull from the `develop` branch, but is good to have in
case we were to change that in the future.

jsf9k committed:
Resolve conflict from follow-on Lineage update

jsf9k committed:
Upgrade to actions/cache@v4 everywhere

Commit:
Merge https://github.com/cisagov/skeleton-generic into lineage/skeleton

# Conflicts:
#	.github/dependabot.yml

jsf9k committed:
Resolve conflict from follow-on Lineage update

jsf9k committed:
Bump up the lower bound on ansible-core

This is being done because the pip-audit pre-commit hook identifies a
vulnerability in ansible-core version 2.16.13.  Note that this
requires that we bump up ansible to version 10 since all versions of
ansible 9 have a dependency on ~=2.16.X.

jsf9k committed:
Bump up the lower bound on ansible-core

This is being done because the pip-audit pre-commit hook identifies a
vulnerability in ansible-core version 2.16.13.  Note that this
requires that we bump up ansible to version 10 since all versions of
ansible 9 have a dependency on ~=2.16.X.

jsf9k committed:
Update the version of the ansible-lint pre-commit hook

Version 24.10.0 is the first version that supports Fedora 41 as a
valid platform.

jsf9k committed:
Adjust pin for ansible-core

The pin of ansible-core was originally put in place because the
pip-audit pre-commit hook identifies a vulnerability in ansible-core
2.16.13.  Normally we would pin ansible-core to >2.16.13, but in the
spirit of the earlier, optional pin of ansible>=10 we pin ansible-core
to >=2.17.  This effectively also pins ansible to >=10.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Upgrade pin for ansible-core

The pin of ansible-core was originally put in place because the
pip-audit pre-commit hook identifies a vulnerability in ansible-core
2.16.13.  Normally we would pin ansible-core accordingly (>2.16.13),
but the earlier pin of ansible>=10 effectively pins ansible-core to
>=2.17 so that's what do.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Ignore a particular ansible-core vulnerability

This is being done only temporarily, and only because there is no
recent version of ansible-core that does not exhibit the
vulnerability.  Without this change we get a failure from the
pip-audit pre-commit hook that we cannot do anything about.

See cisagov/skeleton-packer#380 for more details.

jsf9k committed:
Add comments about looming EOL issues for ansible and ansible-core

This adds even more evidence for why it is a good idea to go ahead and
upgrade ansible and ansible-core, in addition to the vulnerability
that pip-audit turned up.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Add comments about looming EOL issues for ansible and ansible-core

This adds even more evidence for why it is a good idea to go ahead and
upgrade ansible and ansible-core, in addition to the vulnerability
that pip-audit turned up.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Merge pull request #196 from cisagov/improvement/add-a-lower-bound-pin-for-ansible-core

Bump up the lower bound on `ansible-core`

jsf9k committed:
Merge pull request #197 from cisagov/improvement/upgrade-ansible-lint-pre-commit-hook-version

Update the version of the `ansible-lint` `pre-commit` hook

Commit:
Merge https://github.com/cisagov/skeleton-generic into lineage/skeleton
